distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmErgonomics.java

```diff
@@ -0,0 +1,106 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.tools.launchers;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Tunes Elasticsearch JVM settings based on inspection of provided JVM options.
+ */
+final class JvmErgonomics {
+    private static final long KB = 1024L;
+
+    private static final long MB = 1024L * 1024L;
+
+    private static final long GB = 1024L * 1024L * 1024L;
+
+
+    private JvmErgonomics() {
+        throw new AssertionError("No instances intended");
+    }
+
+    /**
+     * Chooses additional JVM options for Elasticsearch.
+     *
+     * @param userDefinedJvmOptions A list of JVM options that have been defined by the user.
+     * @return A list of additional JVM options to set.
+     */
+    static List<String> choose(List<String> userDefinedJvmOptions) {
+        List<String> ergonomicChoices = new ArrayList<>();
+        Long heapSize = extractHeapSize(userDefinedJvmOptions);
+        Map<String, String> systemProperties = extractSystemProperties(userDefinedJvmOptions);
+        if (heapSize != null) {
+            if (heapSize < 1 * GB) {
+                if (systemProperties.containsKey("io.netty.allocator.type") == false) {
+                    ergonomicChoices.add("-Dio.netty.allocator.type=unpooled");
+                }
+            }
+        }
+        return ergonomicChoices;
+    }
+
+    private static final Pattern MAX_HEAP_SIZE = Pattern.compile("^(-Xmx|-XX:MaxHeapSize=)(?<size>\\d+)(?<unit>\\w)?$");
+
+    // package private for testing
+    static Long extractHeapSize(List<String> userDefinedJvmOptions) {
+        for (String jvmOption : userDefinedJvmOptions) {
+            final Matcher matcher = MAX_HEAP_SIZE.matcher(jvmOption);
+            if (matcher.matches()) {
+                final long size = Long.parseLong(matcher.group("size"));
+                final String unit = matcher.group("unit");
+                if (unit == null) {
+                    return size;
+                } else {
+                    switch (unit.toLowerCase(Locale.ROOT)) {
+                        case "k":
+                            return size * KB;
+                        case "m":
+                            return size * MB;
+                        case "g":
+                            return size * GB;
+                        default:
+                            throw new IllegalArgumentException("Unknown unit [" + unit + "] for max heap size in [" + jvmOption + "]");
+                    }
+                }
+            }
+        }
+        return null;
```

Comment:
Would it be possible to return what the default JVM heap is here instead of null? Seems like we still may want to make changes to JVM options for an unset heap.

Either that, or since we explicitly set it in our `jvm.options`, emit a warning or error that no heap has been specified (someone must have removed the option)

Reply:
The JVM chooses the heap size ergonomically when the heap size is not specified.

Reply:
> The JVM chooses the heap size ergonomically when the heap size is not specified.

Do we have the ability to see what it would choose? If running on a small machine, for instance, we'd still want to disable Netty's pooled allocator if the JVM is going to automatically choose a 400mb heap

Reply:
It can be quite tricky because the ergonomics these days depend on other flags on the JVM (e.g., `-XX:+UseCGroupMemoryLimitForHeap`).

Reply:
We could try starting a JVM with all the flags specified and `-XX:+PrintFlagsFinal -version` and scrape the output for `MaxHeapSize`.

Reply:
Thanks for the correction. I've corrected my comment accordingly.

Reply:
> We don't remove the options and use `-XX:+UseCGroupMemoryLimitForHeap` for our docker image(s) do we?

We do not do that.